### PR TITLE
feat(router): add /retrieve api for relay

### PR DESCRIPTION
**Pull request by ShankarSinghC**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

/relay is a api that will be used to just forward the request sent by merchant without performing any application logic on it. This is particular to support the refunds for the payments that was directly performed with the connector (example stripe) with out hyperswitch involved during the payment flow. 

This PR introduces changes to add a retrieve API for relay refunds.

The retrieve API will:

1. Return the relay status from our database when a retrieve call is made.
2. If force_sync = true is passed and the relay status is not in a terminal state in our system, the API will call the connector to fetch the latest status.

Flow
1. Merchant performs payment directly with processor and stores the payment reference id associated with it.
2. Merchant tiggers the refund through hyperswitch using the processor reference id for the payment.
3. Merchant makes retrieve call for the relay status

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

-> Create a merchant connector account
-> Make a refund request with invalid connector_resource_id
```
curl --location 'http://localhost:8080/relay' \
--header 'Content-Type: application/json' \
--header 'X-Profile-Id: pro_rIikb0gPrzd0vEY8hIiP' \
--header 'api-key: dev_qqMQAuMWCclJSk4uMBeNNSpqAPlsKlBA3blQOwOoq5ogTSFVON3taShABqtqYNA4' \
--data '{
  "connector_id": "mca_8W0CQfwE89J1BJBbpcnt",
  "connector_resource_id": "7349526182906968904951",
  "data": {
      "refund": {
    "amount": 50,
    "currency": "USD"
  }},
  "type": "refund"
}'
```
```
{
    "id": "relay_4lgXCQW0EnqKddZ9CC3b",
    "status": "pending",
    "connector_resource_id": "7349526182906968904951",
    "error": null,
    "connector_reference_id": "7349526431306986204951",
    "connector_id": "mca_8W0CQfwE89J1BJBbpcnt",
    "profile_id": "pro_rIikb0gPrzd0vEY8hIiP",
    "type": "refund",
    "data": {
        "refund": {
            "amount": 50,
            "currency": "USD",
            "reason": null
        }
    }
}
```
-> Relay retrieve
```
curl --location 'http://localhost:8080/relay/relay_4lgXCQW0EnqKddZ9CC3b' \
--header 'Content-Type: application/json' \
--header 'X-Idempotency-Key: <x-idempotency-key>' \
--header 'X-Profile-Id: pro_rIikb0gPrzd0vEY8hIiP' \
--header 'api-key: dev_qqMQAuMWCclJSk4uMBeNNSpqAPlsKlBA3blQOwOoq5ogTSFVON3taShABqtqYNA4' \
--data ''
```
```
{
    "id": "relay_4lgXCQW0EnqKddZ9CC3b",
    "status": "pending",
    "connector_resource_id": "7349526182906968904951",
    "error": null,
    "connector_reference_id": "7349526431306986204951",
    "connector_id": "mca_8W0CQfwE89J1BJBbpcnt",
    "profile_id": "pro_rIikb0gPrzd0vEY8hIiP",
    "type": "refund",
    "data": {
        "refund": {
            "amount": 50,
            "currency": "USD",
            "reason": null
        }
    }
}
```

-> Relay retrieve force_sync = `true`
```
curl --location 'http://localhost:8080/relay/relay_4lgXCQW0EnqKddZ9CC3b?force_sync=true' \
--header 'Content-Type: application/json' \
--header 'X-Idempotency-Key: <x-idempotency-key>' \
--header 'X-Profile-Id: pro_rIikb0gPrzd0vEY8hIiP' \
--header 'api-key: dev_qqMQAuMWCclJSk4uMBeNNSpqAPlsKlBA3blQOwOoq5ogTSFVON3taShABqtqYNA4' \
--data ''
```
```
{
    "id": "relay_4lgXCQW0EnqKddZ9CC3b",
    "status": "success",
    "connector_resource_id": "7349526182906968904951",
    "error": null,
    "connector_reference_id": "7349526431306986204951",
    "connector_id": "mca_8W0CQfwE89J1BJBbpcnt",
    "profile_id": "pro_rIikb0gPrzd0vEY8hIiP",
    "type": "refund",
    "data": {
        "refund": {
            "amount": 50,
            "currency": "USD",
            "reason": null
        }
    }
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/relay.rs [hunk 1]
```diff
@@ -175,3 +176,162 @@ pub fn validate_relay_refund_data(
...
+    req: relay_models::RelayRetrieveRequest,
+) -> RouterResponse<relay_models::RelayResponse> {
+    let db = state.store.as_ref();
+    let key_manager_state = &(&state).into();
+    let merchant_id = merchant_account.get_id();
+    let relay_id = &req.id;
+
+    let profile_id_from_auth_layer = profile_id_optional
+        .get_required_value("ProfileId")
+        .change_context(errors::ApiErrorResponse::MissingRequiredField {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/relay.rs:194]: need not do the change_context here

File: crates/diesel_models/src/query/relay.rs [hunk 2]
```diff
@@ -35,4 +35,18 @@ impl Relay {
...
     }
+
+    pub async fn find_by_id_merchant_id(
+        conn: &PgPooledConn,
+        merchant_id: &common_utils::id_type::MerchantId,
+        id: &common_utils::id_type::RelayId,
+    ) -> StorageResult<Self> {
+        generics::generic_find_one::<<Self as HasTable>::Table, _, _>(
+            conn,
+            dsl::merchant_id
```
Comment (sai-harsha-vardhan) [inline crates/diesel_models/src/query/relay.rs:46]: nit: consider adding INDEX for (id, merchant_id)

File: crates/router/src/core/relay.rs [hunk 3]
```diff
@@ -175,3 +176,162 @@ pub fn validate_relay_refund_data(
...
+        response,
+    ))
+}
+
+fn should_call_connector_for_relay_refund_status(
+    relay: &hyperswitch_domain_models::relay::Relay,
+    force_sync: bool,
+) -> bool {
+    // This allows refund sync at connector level if force_sync is enabled, or
+    // checks if the refund has failed
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/relay.rs:286]: nit: checkf if the refund is in terminal state

File: crates/router/src/routes/relay.rs [hunk 4]
```diff
@@ -38,3 +38,39 @@ pub async fn relay(
...
+
+#[instrument(skip_all, fields(flow = ?Flow::Relay))]
+#[cfg(feature = "oltp")]
+pub async fn relay_retrieve(
+    state: web::Data<app::AppState>,
+    path: web::Path<common_utils::id_type::RelayId>,
+    req: actix_web::HttpRequest,
+    query_params: web::Query<api_models::relay::RelayRetrieveBody>,
+) -> impl Responder {
+    let flow = Flow::Relay;
```
Comment (sai-harsha-vardhan) [inline crates/router/src/routes/relay.rs:50]:
```suggestion
    let flow = Flow::RelayRetrieve;
```

File: crates/router/src/core/relay.rs [hunk 5]
```diff
@@ -175,3 +173,158 @@ pub fn validate_relay_refund_data(
...
+        merchant_id,
+        &profile_id_from_auth_layer,
+    )
+    .await
+    .change_context(errors::ApiErrorResponse::ProfileNotFound {
+        id: profile_id_from_auth_layer.get_string_repr().to_owned(),
+    })?;
+
+    let relay_record_result = db
+        .find_relay_by_id_merchant_id(key_manager_state, &key_store, relay_id)
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/relay.rs:203]:
```suggestion
        .find_relay_by_id(key_manager_state, &key_store, relay_id)
```

**Review by jarnura — DISMISSED** [review]

File: crates/router/src/core/relay.rs (hunk 5 above)
Comment (jarnura) [reply crates/router/src/core/relay.rs:203]: +1

**Review by sai-harsha-vardhan — APPROVED** [review]

**Pull request merged**